Metrics.py

```diff
@@ -37,16 +47,323 @@ def truePosOrNeg(self, dataSet, truePosOrNeg):
+			TNTotal += trueNegCount
+
+		keys = truePosByAttribute.keys()
+		chiSquareVal = self.chiSquare(truePosByAttribute, trueNegByAttribute, TPTotal, TNTotal)
+
+		# to find degreeOfFreedom, multiply rows - 1 by columns - 1, num rows is always 2, thus this is just columns - 1
+		degreeOfFreedom = len(keys) - 1
+
+		# finds p value using the cumulative distribution function
+		pValue = 1 - stats.chi2.cdf(chiSquareVal, degreeOfFreedom)
+
+		EquOfOpp = False
+		if pValue < 0.05:
+			EquOfOpp = True
+
+		return pValue, EquOfOpp
+
+	# TODO: train vs test part of dataset? help?
+	'''
+	Computes the fairness of trainedBayes according to counterfactual measures by running Bayes on the original data, then
+	swapping the protected attribute values and reclassifying (without retraining). Then, to compute accuracy, it treats the
+	original classifications as if they were true.
+		dataSet (DataSet) - the original dataset
+		trainedBayes (Bayes object) - the trained Bayes model
+		
+	return: accuracy
+	'''
+
+	# TODO: Test this
+	def counterfactualMeasures(self, dataSet, trainedBayes):
+		swappedDataSet = self.swapProtectedAttributes(dataSet)
+		swappedDF = swappedDataSet.dataFrame
+		swappedDF.drop(columns=["Bayes Classification", swappedDataSet.trueLabels])
+
+		trainedBayes.classify(swappedDataSet)
+		swappedDF = swappedDataSet.dataFrame
+
+		swappedDF[swappedDataSet.trueLabels] = dataSet.dataFrame["Bayes Classification"].astype(int)
+
+		return self.calculateAccuracy(swappedDataSet)
+
+	'''
+	Copies the dataset, then swaps the protected attribute values in the copy.
+		dataSet (DataSet) - the original dataset
+		
+	returns: a copy of dataSet with the protected attribute values swapped.
+	'''
+	def swapProtectedAttributes(self, dataSet):
+		dataSetCopy = dataSet.copyDataSet()
+		dataFrame = dataSetCopy.dataFrame
+
+		possibleAttributeValues = dataFrame[dataSetCopy.protectedAttribute].unique()
+
+		if len(possibleAttributeValues) != 2:
+			return "Cannot calculate counterfactual measures for nonbinary protected attributes."
+
+		for i in range(dataFrame.shape[0]):
+			protectedAttributeValue = dataFrame.at[i, dataSetCopy.protectedAttribute]
+			if protectedAttributeValue == possibleAttributeValues[0]:
+				dataFrame.loc[[i], [dataSetCopy.protectedAttribute]] = possibleAttributeValues[1]
+
+			else:
+				dataFrame.loc[[i], [dataSetCopy.protectedAttribute]] = possibleAttributeValues[0]
+
+		return dataSetCopy
+
+	'''
+	Counts the total number of preferred outcomes for a particular protected attribute class 
+		dataset (DataSet) - the original dataset
+		
+	returns: a dictionary where the keys are protected attribute values and the values are the number of positive
+		outcomes that protected attribute group received.
+	'''
+	def countPositiveOutcomes(self, dataSet):
+		dataFrame = dataSet.dataFrame
+		possibleAttributes = dataFrame[dataSet.protectedAttributes[0]].unique()
+
+		posOutcomes0 = 0
+		posOutcomes1 = 0
```

Comment:
Changing the names of these variables or explaining what they hold in the function comment may be useful for readability.